f5_openstack_agent/utils/cert_parser.py

```diff
@@ -0,0 +1,208 @@
+# coding=utf-8
+# Copyright (c) 2016-2018, F5 Networks, Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import base64
+
+from cryptography.hazmat import backends
+from cryptography.hazmat.primitives import serialization
+from cryptography import x509
+from f5_openstack_agent.lbaasv2.drivers.bigip import exceptions as f5_ex
+from oslo_log import log as logging
+from pyasn1.codec.der import decoder as der_decoder
+from pyasn1.codec.der import encoder as der_encoder
+from pyasn1_modules import rfc2315
+import six
```

Comment:
Are these modules included as dependencies by default or expected to be installed as part of the env? Is it the convention to assume these 3rd party libs will be installed separately from the plugin? Or would they need to be included in packaging ref such as `setup_requirements.txt` or `setup.py`?

Reply:
I checked the modules, they are in Openstack global requirements https://github.com/openstack/requirements. 
I think it is better to write in the setup_requirements.txt

Reply:
It seems the test won pass when I set the setup_requirements.txt and requirements.unittest.txt file.

I pushed three times.
The first time, I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1>=0.1.9,<=0.4.3
       pyasn1-modules>=0.0.8,<=0.2.2
       cryptography>=1.2.3,<=2.3

error:
`dpkg-gencontrol: warning: can't parse dependency python-pyasn1 (<=0.4.3
dpkg-gencontrol: error: error occurred while parsing Depends field: , python (>= 2.7), python (<< 2.8), python:any (>= 2.7.1-0ubuntu2), python:any, python-f5-sdk, python-f5-sdk (<< 3.0.12), python-f5-sdk (>= 3.0.11), python-six (>=1.9.0), python-pyasn1 (<=0.4.3, >=0.1.9), python-pyasn1-modules (<=0.2.2, >=0.0.8), python-cryptography (<=2.3, >=1.2.3)`


The second time, I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1<=0.4.3
       pyasn1-modules<=0.2.2
       cryptography<=2.3

error:
`Can't get package dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm
 -- rpm -ivh /tmp/*.rpm
Agent install failed
The command "f5-openstack-agent-dist/scripts/test_install.sh "redhat" "7" ${PKG_RELEASE_EL7}" exited with 1.`


The third time (this time), I add below content to requirements.unittest.txt and  setup_requirements.txt:
       six>=1.9.0 # MIT
       pyasn1>=0.1.9
       pyasn1-modules>=0.0.8
       cryptography>=1.2.3

error:
`Failed to copy f5-openstack-agent package
Getting dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm.
 -- rpm -qRp f5-openstack-agent-dist/rpms/build/f5-openstack-agent-10.2.0-1.el7.noarch.rpm
Can't get package dependencies for f5-openstack-agent-10.2.0-1.el7.noarch.rpm
 -- rpm -ivh /tmp/*.rpm
Agent install failed
The command "f5-openstack-agent-dist/scripts/test_install.sh "redhat" "7" ${PKG_RELEASE_EL7}" exited with 1.`

`Installing: /tmp/python-f5-sdk-rest_3.0.11_1404_all.deb
SDK Succeeded in install test
Installing: /tmp/python-f5-openstack-agent_10.2.0-1_1404_all.deb
SDK install failed (1)
The command "f5-openstack-agent-dist/scripts/test_install.sh "ubuntu" "14.04" ${PKG_RELEASE_1404}" exited with 1.`

I am not very sure why this happened. Do you have any idea to check the process in detail or why did it happen?

Reply:
@kaustriaf5

Reply:
The problem happens when it install the packages on Redhat using rpm and Ubuntu using dpkg.